### DISCUSSION TITLE
[@types/webextension-polyfill] Property 'declarativeContent' does not exist on type 'typeof Browser'

### DISCUSSION THREAD

**Discussion by dan-lovelace** [DefinitelyTyped/DefinitelyTyped#66395] [Issues with a @types package]
After upgrading `@types/webextension-polyfill` to version `0.10.1`, I started getting the error `Property 'declarativeContent' does not exist on type 'typeof Browser'`. After doing some digging, I can see this change was introduced in #63847 here: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/63847/files#diff-5c6517863f167a7879c2af96cf121f4cef3fb85d7c66e56e401e790979165d55L22.

Here's a usage example taken from the [Chrome developer docs](https://developer.chrome.com/docs/extensions/reference/action/#emulating-pageactions-with-declarativecontent):

```js
import browser from "webextension-polyfill";

browser.declarativeContent.onPageChanged.removeRules(undefined, () => {
  browser.declarativeContent.onPageChanged.addRules([
    {
      conditions: [
        new browser.declarativeContent.PageStateMatcher({
          pageUrl: { hostSuffix: "github.com" },
        }),
      ],
      actions: [new browser.declarativeContent.ShowAction()],
    },
  ]);
});
```

I'm wondering if removing support for this was intended and if so, what alternatives are available. Looking at the changes, it was replaced with `declarativeNetRequest` which is a completely unrelated API. For now, the solution appears to be to keep using the older version `0.9.1`.

[a comment by an automated account is not shown]

**Comment by Lusito** [comment]
Hi @dan-lovelace,

sorry for the delay. That is a very good question. The changes introduced where caused by the upstream schema changes by mozilla. They actively removed it themselves and my code just autogenerated the declarations.

I've digged a bit and found this issue: https://bugzilla.mozilla.org/show_bug.cgi?id=1435864

The relevant comment is:
> The development of declarativeContent is currently on hold because the risk of the implementation does not balance against the potential benefits of it. For more details, see https://docs.google.com/document/d/1KjC2uAV0n5BszbyQjclzp2cLiOhWuKrTmqvqyz5ggKU/preview
>
> There are several alternatives to the declarativeContent API, such as the tabs and webNavigation APIs, and using them results in code that is still compatible with Chrome.

Since this is about the webextension polyfill and not about pure-chrome definitions, I'd say it's better to use the cross-browser APIs.
Is there any reason you can't use those other APIs?

**Reply by dan-lovelace** [reply]
Hey @Lusito thanks for getting back. I've since moved away from the `declarativeContent` API. I was originally using it to enable/disable the extension's icon on certain domains but I found it to be a poor user experience. Chrome recommends against it as well somewhere in their documentation.

I'll close this one out! Thanks again.

**Reply by Lusito** [reply]
I'll have to correct my statement above. The file was actually fetched from chrome schemas, but they changed their branch name from master to main and my generator script swallowed the error. I've corrected that bug, so that bad fetches will break the build, but am wondering if it still should be removed. Can you link me to the chrome page, where it's recommended not to use it?

**Reply by dan-lovelace** [reply]
I think you should keep support for `declarativeContent` since it is actively documented in the Chrome documentation: https://developer.chrome.com/docs/extensions/reference/declarativeContent/. They even have an example for using it to emulate the `pageAction` API: https://developer.chrome.com/docs/extensions/reference/action/#emulating-pageactions-with-declarativecontent. This is where I initially got the idea of using it to en/disable the icon.

My previous statement about them recommending against usage was unclear. They don't specifically say to _not_ use `declarativeContent`, they only mention avoiding constant animation of the page action: https://developer.chrome.com/docs/extensions/reference/pageAction/#tips. I took this suggestion for my own use case and decided to always allow my users to open the popup window.
